packages/wallet/ui/src/bridge-dapp.jsx

```diff
@@ -0,0 +1,190 @@
+  if (me === parent) {
+    throw Error('window.parent === parent!!!');
+  }
+};
+
+/**
+ * Requests a dapp at a given origin to connect to the wallet UI. This will
+ * prompt the user to accept the dapp in the wallet and allow it to propose
+ * offers if accepted.
+ *
+ * @param {DappKey} dappKey
+ * @param {string} proposedPetname - The suggested petname if the wallet does
+ * not already know about the dapp.
+ */
+const requestDappConnection = (dappKey, proposedPetname) => {
+  const dapp = loadDapp(dappKey);
+  if (dapp) {
+    return;
+  }
+  upsertDapp(dappKey.chainId, dappKey.address, {
+    origin: dappKey.origin,
+    petname: proposedPetname,
+  });
+};
+
+/**
+ * Watches for changes in local storage to the dapp's approval status and
+ * notifies the dapp.
+ *
+ * @param {DappKey} dappKey
+ * @param {boolean} currentlyApproved
+ */
+const watchDappApproval = (dappKey, currentlyApproved) => {
+  watchDapps(dappKey.chainId, dappKey.address, dapps => {
+    const dapp = dapps.find(d => d.origin === dappKey.origin);
+    const isDappApproved = dapp && dapp.isEnabled;
+    if (isDappApproved !== currentlyApproved) {
+      sendMessage({
+        type: BridgeProtocol.dappApprovalChanged,
+        isDappApproved,
+      });
+      currentlyApproved = isDappApproved;
+    }
+  });
+};
+
+/** @typedef {import('@agoric/web-components/src/dapp-wallet-bridge/DappWalletBridge').OfferConfig} OfferConfig */
+/**
+ * Propose an offer from the dapp to the wallet UI.
+ *
+ * @param {DappKey} dappKey
+ * @param {OfferConfig} offerConfig
+ */
+const createAndAddOffer = (dappKey, offerConfig) => {
+  const dapp = loadDapp(dappKey);
+  const isDappApproved = !!dapp?.isEnabled;
+  if (!isDappApproved) return;
+
+  const currentTime = new Date().getTime();
+  // TODO(https://github.com/Agoric/agoric-sdk/issues/6478): Generate separate
+  // id for transaction when signing.
+  const id = currentTime;
+
+  const offer = {
+    id,
+    meta: {
+      id,
+      creationStamp: currentTime,
+    },
+    requestContext: { origin: dappKey.origin },
+    status: OfferUIStatus.proposed,
+    ...offerConfig,
+  };
+  addOffer(dappKey.chainId, dappKey.address, offer);
+};
+
+/**
+ * Notifies the dapp about whether it's approved or not and watches for changes
+ * to its approval status.
+ *
+ * @param {DappKey} dappKey
+ */
+const checkAndWatchDappApproval = dappKey => {
+  // Dapps are keyed by origin. A dapp is basically an origin with a petname
+  // and an approval status.
+  const dapp = loadDapp(dappKey);
+  const isDappApproved = !!dapp?.isEnabled;
+  sendMessage({
+    type: BridgeProtocol.checkIfDappApproved,
+    isDappApproved,
+  });
+  watchDappApproval(dappKey, isDappApproved);
+};
+
+const handleIncomingMessages = () => {
+  /** @type {DappKey=} */
+  let dappKey;
+
+  window.addEventListener('message', ev => {
+    const type = ev.data?.type;
+    if (!type?.startsWith(BridgeProtocol.prefix)) {
+      return;
+    }
+    if (dappKey === undefined) {
```

Comment:
IIUC this is only supposed to happen for a `checkIfDappApproved` which must be the first message. If so, please move this under that case.

or… add a stronger check here like,
```
assert.same(type, BridgeProtocol.checkIfDappApproved, 'first message must be checkIfDappApproved');
```

Reply:
Done